File: Packs/CommonScripts/Scripts/GeneratePassword/GeneratePassword.js
dArgs = {
    'debug': args.debug = (args.debug === 'true'),
    'max_lcase': parseInt(args.max_lcase),
    'min_lcase': parseInt(args.min_lcase),
    'max_ucase': parseInt(args.max_ucase),
    'min_ucase': parseInt(args.min_lcase),
    'max_digits': parseInt(args.max_digits),
    'min_digits': parseInt(args.min_digits),
    'max_symbols': parseInt(args.max_symbols),
    'min_symbols': parseInt(args.min_symbols)
};

// randomize our selected charaters
function randomSort(a, b) {
    return Math.random() > 0.5 ? -1 : 1;
}

function printCharValues(pw) {
    var s = [];
    for (i = 0; i < pw.length; i++) {
        s.push(pw.charCodeAt(i));
    }
    log("Ascii for password = " + s.toString());
}

if(args.debug) {
   log(JSON.stringify(dArgs));
}

// Define the characters of our classes
var lcase   = "abcdefghijklmnopqrstuvwxyz";
var ucase   = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
var n       = "0123456789";
var s       = "!@#$%^&*()[]+:\"?_><=';/-.,\\|";

// randomize the amount of characters we get as per parameters
var numu = dArgs.max_ucase   - dArgs.min_ucase   >= 0 ? Math.floor(Math.random() * (dArgs.max_ucase   - dArgs.min_ucase   + 1)) + dArgs.min_ucase   : 0;
var numl = dArgs.max_lcase   - dArgs.min_lcase   >= 0 ? Math.floor(Math.random() * (dArgs.max_lcase   - dArgs.min_lcase   + 1)) + dArgs.min_lcase   : 0;
var numn = dArgs.max_digits  - dArgs.min_digits  >= 0 ? Math.floor(Math.random() * (dArgs.max_digits  - dArgs.min_digits  + 1)) + dArgs.min_digits  : 0;
var nums = dArgs.max_symbols - dArgs.min_symbols >= 0 ? Math.floor(Math.random() * (dArgs.max_symbols - dArgs.min_symbols + 1)) + dArgs.min_symbols : 0;

if(numu + numl + numn + nums === 0) {
   return  { ContentsFormat: formats.text, Type:entryTypes.error, Contents: 'error: insane password. No character length.'};
}

// start with a blank password.
var pw = "";

// iterate through each character class and add
for (var i = 0; i < numu; i++) {
   pw += ucase[Math.floor(Math.random() * ucase.length)];
}
for (var i = 0; i < numl; i++) {
   pw += lcase[Math.floor(Math.random() * lcase.length)];
}
for (var i = 0; i < numn; i++) {
   pw += n[Math.floor(Math.random() * n.length)];
}
for (var i = 0; i < nums; i++) {
   pw += s[Math.floor(Math.random() * s.length)];
}

// randomize our new password string
var rpw = (pw.split('').sort(randomSort)).join('');

if(args.debug) {
    printCharValues(rpw);
}
return {
        Type: entryTypes.note,
        Contents: {"NEW_PASSWORD": rpw},          // used by raw_contents = true
        ContentsFormat: formats.json,             // defines the source format
        HumanReadable: tableToMarkdown('Newly Generated Password', {"password": rpw}),
        EntryContext: {"NEW_PASSWORD": rpw}       // same as setcontext
};

File: Packs/CommonScripts/Scripts/GeneratePassword/GeneratePassword.py
import demistomock as demisto
from CommonServerPython import *
from CommonServerUserPython import *

import random
import secrets
import string

DEFAULT_MIN = 0
DEFAULT_MAX = 10

LOWERCASE_CHARS = string.ascii_lowercase
UPPERCASE_CHARS = string.ascii_uppercase
DIGITS = string.digits
SYMBOLS = "!@#$%^&*()[]+:\"?_><=';/-.,\\|"


def randomize_number_in_range(min_val: int, max_val: int) -> int:
    n = max(max_val - min_val, 0)
    return secrets.randbelow(n + 1) + min_val


def print_char_values(pw):
    ascii_values = [ord(char) for char in pw]
    ascii_string = ', '.join(str(value) for value in ascii_values)
    demisto.debug(f"ASCII for password = {ascii_string}")


def generate_password(args: Dict[str, Any]) -> CommandResults:
    is_debug = argToBoolean(args.get('debug'))
    min_lowercase = arg_to_number(args.get('min_lcase', DEFAULT_MIN))
    max_lowercase = arg_to_number(args.get('max_lcase', DEFAULT_MAX))
    min_uppercase = arg_to_number(args.get('min_ucase', DEFAULT_MIN))
    max_uppercase = arg_to_number(args.get('max_ucase', DEFAULT_MAX))
    min_digits = arg_to_number(args.get('min_digits', DEFAULT_MIN))
    max_digits = arg_to_number(args.get('max_digits', DEFAULT_MAX))
    min_symbols = arg_to_number(args.get('min_symbols', DEFAULT_MIN))
    max_symbols = arg_to_number(args.get('max_symbols', DEFAULT_MAX))

    if min(min_uppercase, min_lowercase, min_digits, min_symbols) < 0:  # type:ignore[type-var,operator]
        raise DemistoException("All numeral arguments must be positive.")

    if sum((min_uppercase, min_lowercase, min_digits, min_symbols)) == 0:  # type:ignore[arg-type]
        raise DemistoException("error: At least one of the following arguments should be above 0: "
                               "min_uppercase, min_lowercase, min_digits, min_symbols")

    # randomize the amount of characters we get as per parameters
    num_upper = randomize_number_in_range(min_uppercase, max_uppercase)  # type:ignore[arg-type]
    num_lower = randomize_number_in_range(min_lowercase, max_lowercase)  # type:ignore[arg-type]
    num_digits = randomize_number_in_range(min_digits, max_digits)  # type:ignore[arg-type]
    num_symbols = randomize_number_in_range(min_symbols, max_symbols)  # type:ignore[arg-type]

    if num_upper + num_lower + num_digits + num_symbols == 0:
        raise DemistoException("error: insane password. No character length.")

    # start with a blank password
    pw = []

    # iterate through each character class and add
    for _ in range(num_upper):
        pw.append(secrets.choice(UPPERCASE_CHARS))
    for _ in range(num_lower):
        pw.append(secrets.choice(LOWERCASE_CHARS))
    for _ in range(num_digits):
        pw.append(secrets.choice(DIGITS))
    for _ in range(num_symbols):
        pw.append(secrets.choice(SYMBOLS))

    # randomize our new password string
    rpw = ''.join(random.sample(pw, len(pw)))

    if is_debug:
        print_char_values(rpw)

    return CommandResults(
        outputs_prefix="NEW_PASSWORD",
        outputs=rpw,
        readable_output=tableToMarkdown('Newly Generated Password', {'password': rpw})
    )


def main():  # pragma: no cover
    try:
        args = demisto.args()
        return_results(generate_password(args))
    except Exception as e:
        return_error(str(e))


if __name__ in ('__builtin__', 'builtins'):  # pragma: no cover
    main()
